pkg/cmd/roachtest/tests/mixed_version_cdc.go

```diff
@@ -293,122 +349,97 @@ func (cmvt *cdcMixedVersionTester) timestampResolved() {
+			Option("tolerate-errors")
+
+		// Sanity check that the workload has been initlized.
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-cmvt.workloadInit:
+		}
 
-	// sqlNode returns the node to be used when sending SQL statements
-	// during this test. It is randomized, but the random seed is logged
-	// above.
-	sqlNode := func() int {
-		return tester.crdbNodes[rng.Intn(len(tester.crdbNodes))]
+		h.BackgroundCommand(bankRun.String(), option.NodeListOption{h.RandomNode(r, cmvt.crdbNodes)})
+		return nil
 	}
+}
 
-	newVersionUpgradeTest(c,
-		uploadAndStartFromCheckpointFixture(tester.crdbNodes, predecessorVersion),
-		tester.setupVerifier(sqlNode()),
-		tester.installAndStartWorkload(),
-		waitForUpgradeStep(tester.crdbNodes),
-
-		// NB: at this point, cluster and binary version equal predecessorVersion,
-		// and auto-upgrades are on.
-		preventAutoUpgradeStep(sqlNode()),
-		tester.createChangeFeed(sqlNode()),
-
-		tester.waitForResolvedTimestamps(),
-		// Roll the nodes into the new version one by one in random order
-		tester.crdbUpgradeStep(binaryUpgradeStep(tester.crdbNodes, clusterupgrade.MainVersion)),
-		// let the workload run in the new version for a while
-		tester.waitForResolvedTimestamps(),
-
-		tester.assertValid(),
-
-		// Roll back again, which ought to be fine because the cluster upgrade was
-		// not finalized.
-		tester.crdbUpgradeStep(binaryUpgradeStep(tester.crdbNodes, predecessorVersion)),
-		tester.waitForResolvedTimestamps(),
-
-		tester.assertValid(),
+// initWorkload synchronously initializes the workload.
+func (cmvt *cdcMixedVersionTester) initWorkload() func(context.Context, *logger.Logger, *rand.Rand, *mixedversion.Helper) error {
+	return func(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
+		bankInit := roachtestutil.NewCommand("%s workload init bank", test.DefaultCockroachPath).
+			Arg("{pgurl:1}")
 
-		// Roll nodes forward and finalize upgrade.
-		tester.crdbUpgradeStep(binaryUpgradeStep(tester.crdbNodes, clusterupgrade.MainVersion)),
+		// NB: We add 1 because option.NodeListOptions are indexed starting from 1.
+		if err := h.Command(bankInit.String(), option.NodeListOption{h.RandomNode(r, cmvt.crdbNodes) + 1}); err != nil {
+			return err
+		}
+		close(cmvt.workloadInit)
+		return nil
+	}
+}
 
-		// allow cluster version to update
-		allowAutoUpgradeStep(sqlNode()),
-		waitForUpgradeStep(tester.crdbNodes),
+func runCDCMixedVersions(ctx context.Context, t test.Test, c cluster.Cluster) {
+	tester := newCDCMixedVersionTester(ctx, t, c)
 
-		tester.waitForResolvedTimestamps(),
-		tester.finishTest(),
-		tester.assertValid(),
-	).run(ctx, t)
+	// NB: We rely on the testing framework to choose a random predecessor
+	// to upgrade from.
+	mvt := mixedversion.NewTest(ctx, t, t.L(), c, tester.crdbNodes)
+
+	cleanupKafka := tester.StartKafka(t, c)
+	defer cleanupKafka()
+
+	// Register hooks.
+	mvt.OnStartup("start-changefeed", tester.createChangeFeed())
```

Comment:
Nit: I'd use human readable descriptions for these functions (i.e., `start changefeed` instead of `start-changefeed`) -- that first string is for human consumption. Log files already sanitize the string to make them filename-safe.